src/Tools/SchemaTool.php

```diff
@@ -969,29 +971,84 @@ private function createSchemaForComparison(Schema $toSchema): Schema
-            }
+        return $table->getPrimaryKeyConstraint();
+    }
 
-            $table->addPrimaryKeyConstraint(new PrimaryKeyConstraint(null, $primaryKeyColumnNames, true));
-        } else {
+    /** @param string[] $primaryKeyColumns */
+    private function addPrimaryKeyConstraint(Table $table, array $primaryKeyColumns): PrimaryKeyConstraint|Index
+    {
+        // DBAL < 4.3
+        if (! class_exists(PrimaryKeyConstraint::class)) {
             $table->setPrimaryKey($primaryKeyColumns);
+
+            return $table->getPrimaryKey();
         }
+
+        $primaryKeyConstraint = PrimaryKeyConstraint::editor()
+            ->setUnquotedColumnNames(...array_values($primaryKeyColumns))
+            ->setIsClustered(true)
+            ->create();
+
+        $table->addPrimaryKeyConstraint($primaryKeyConstraint);
+
+        return $table->getPrimaryKeyConstraint();
     }
 
-    /** @return string[] */
+    /** @return non-empty-list<string> */
     private static function getIndexedColumns(Index $index): array
     {
-        if (method_exists(Index::class, 'getIndexedColumns')) {
-            return array_map(static fn (IndexedColumn $indexedColumn) => $indexedColumn->getColumnName()->toString(), $index->getIndexedColumns());
+        // DBAL < 4.3
+        if (! method_exists(Index::class, 'getIndexedColumns')) {
+            return $index->getColumns();
+        }
+
+        return array_map(static fn (IndexedColumn $indexedColumn) => $indexedColumn->getColumnName()->getIdentifier()->getValue(), $index->getIndexedColumns());
+    }
+
+    private function doesIndexOverlapWithPrimaryKey(Index $index, PrimaryKeyConstraint|Index $primaryKey): bool
+    {
+        // DBAL < 4.3
+        if ($primaryKey instanceof Index) {
+            return $index !== $primaryKey && $primaryKey->spansColumns(self::getIndexedColumns($index));
+        }
+
+        $indexedColumns = $index->getIndexedColumns();
+        foreach ($primaryKey->getColumnNames() as $i => $column) {
+            if (
+                ! isset($indexedColumns[$i])
+                || strtolower($column->getIdentifier()->getValue())
+                    !== strtolower($indexedColumns[$i]->getColumnName()->getIdentifier()->getValue())
+            ) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @param non-empty-string                 $indexName
+     * @param non-empty-list<non-empty-string> $indexColumns
+     * @param array<string, mixed>             $indexOptions
+     */
+    private function createIndexForComparison(string $indexName, array $indexColumns, mixed $indexOptions): Index
+    {
+        // DBAL < 4.3
+        if (! class_exists(IndexEditor::class)) {
+            return new Index($indexName, $indexColumns, true, false, [], $indexOptions);
         }
 
-        return $index->getColumns();
+        return Index::editor()
+            ->setUnquotedName($indexName)
```

Comment:
I would double-check how this code behaves if the index name contains quotes (e.g. `"123_index"`). Previously, `Index` would parse it resulting in quoted name `123_index`. In the new version, it will result in an unquoted name `"123_index"` (i.e. quotes will be included in the name, not part of the syntax).

If that's the case, you'll need to use parse the string with `UnqualifiedNameParser` and use `setName()` instead of `setUnquotedName()`.